release-20.2: sql,catalog/descs,gcjob,lease,importccl,backupccl: replace PublishMultiple

ajwerner opened the pull request:
Backport 12/12 commits fro54m #53727.

/cc @cockroachdb/release

---

This series of commits removed `lease.Manager.PublishMultiple()` and its variants and replaces them with a new transactional wrapper around the `descs.Collection`.

Fixes #50906
Fixes #51417

----

Backport 1/1 commits from #54640

Previously, in `descs.Txn()`, we weren't clearing the state of the
associated `descs.Collection` by calling `ReleaseAll()` when the
transaction hit a retry error upon attempting to commit. Consequently,
descriptors we had attempted to write would persist in the collection's
set of uncommitted descriptors on the next attempt. This manifested in
flakes where import jobs would fail because they expected to be reading
a descriptor from the store, but instead read one with a higher version.
To fix this, we unconditionally `defer` the call to `ReleaseAll()` in
the transaction closure.

Fixes #51197.
Fixes #54630.
Fixes #54549.
Fixes #54385.

Commit:
sql: add commentary around leasing of descriptors during backfill

Release justification: non-production code changes

Release note: None

Commit:
sql: clean up backfill internal planner cleanup

Bit of a drive-by.

Release justification: bug fixes and low-risk updates to new functionality

Release note: None

Commit:
sql,catalog/descs: extract CheckTwoVersionInvariant, add descs.Txn

Release justification: low risk, high benefit changes to existing functionality
Release note: None

Commit:
sql: adopt descs.Txn in type and schema changer

This commit removes all uses of `lease.Publish` and `lease.PublishMultiple`

Release justification: low risk, high benefit change to existing functionality.

Release note: None

Commit:
catalogkv: extract a function to write namespace entry removal to a batch

Release justification: low-risk updates to new functionality

Release note: None

Commit:
sql,catalog/descs: add descs.Collection.WriteDescToBatch

This method is handy as all over the place we want to consider incrementing the
version, add the uncommitted descriptor to the `Collection` and then write it
to a batch. We use this more later.

It also permits adding an uncommitted descriptor which is new. This enables
adding a new descriptor that has a version other than 1.

Release justification: low-risk updates to new functionality.

Release note: None

Commit:
catalog/descs: export GetMutableDescriptorByID

Release justification: low-risk updates to new functionality

Release note: None

Commit:
gcjob: replace lease.Manager.Publish with descs.Txn

Release justification: low risk, high benefit changes to existing functionality

Release note: None

Commit:
catalog/lease: move Publish and PublishMultiple to testing

Release justification: non-production code changes.

Release note: None

Commit:
importccl,backupccl: adopt descs.Txn and general descriptor handling

This commit adopts descs.Txn and the relevant new descs.Collection methods in
restore. It should eliminate all of the cases where we might have written a
mutated descriptor that was not read from the store. It also eliminates the
conditional reading of descriptors which in the later step was deemed to be
somewhat paranoid. In the initial transaction of the two jobs it makes a
version mismatch into an error.

Release justification: low risk, high benefit changes to existing functionality

Release note: None

Commit:
tabledesc: don't increment the version in MaybeIncrementVersion with Created

This causes problems. In the other descriptors we use a pointer for the
cluster version and then check on that nil in `MaybeIncrementVersion`. For
legacy reasons, it is not a pointer in the `tabledesc.Mutable`. Nevertheless,
make the logic match.

Release justification: bug fixes and low-risk updates to new functionality

Release note: None

Commit:
catalog/descs: deal somewhat with offline states resolving user-defined types

This commit ensures that offline schemas can be resolved in the context of
resolving mutable table descriptors which utilize user-defined types.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54495)
<!-- Reviewable:end -->

Commit:
sql/catalog: fix descs.Collection reuse on retries in descs.Txn

Previously, in `descs.Txn()`, we weren't clearing the state of the
associated `descs.Collection` by calling `ReleaseAll()` when the
transaction hit a retry error upon attempting to commit. Consequently,
descriptors we had attempted to write would persist in the collection's
set of uncommitted descriptors on the next attempt. This manifested in
flakes where import jobs would fail because they expected to be reading
a descriptor from the store, but instead read one with a higher version.
To fix this, we unconditionally `defer` the call to `ReleaseAll()` in
the transaction closure.

Release note: None (bug has not been released)

ajwerner commented:
@lucy-zhang I cherry-picked the commit from https://github.com/cockroachdb/cockroach/pull/54640 onto this.
